x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java

```diff
@@ -232,22 +219,32 @@ public List<Gap> waitForRange(final ByteRange range, final ByteRange subRange, f
             assert ranges.containsAll(pendingRanges) : ranges + " vs " + pendingRanges;
             assert pendingRanges.stream().allMatch(Range::isPending) : pendingRanges;
             assert pendingRanges.size() != 1 || gaps.size() <= 1 : gaps;
+        }
 
-            // Pending ranges that needs to be filled before executing the listener
-            requiredRanges = range.equals(subRange)
-                ? pendingRanges
-                : pendingRanges.stream()
-                    .filter(pendingRange -> pendingRange.start < subRange.end())
-                    .filter(pendingRange -> subRange.start() < pendingRange.end)
-                    .sorted(Comparator.comparingLong(r -> r.start))
-                    .collect(Collectors.toList());
+        // Pending ranges that needs to be filled before executing the listener
+        if (range.equals(subRange) == false) {
```

Comment:
Is it a referential equality check? it seems that [Range](https://github.com/elastic/elasticsearch/blob/6da721eddab2d9f7c1b51ff5aba7da25634ab5ac/x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java#L565) missed `equals` override.

Reply:
This is a `org.elasticsearch.blobcache.common.ByteRange` which has a proper equals, not a `Range`.